RetainPtr<>::leakRef() should have NS/CF_RETURNS_RETAINED attributes

ddkilzer opened the pull request:
#### 611c1a1ceeaf4a530810d3d60c7460ff89bf1767
<pre>
RetainPtr&lt;&gt;::leakRef() should have NS/CF_RETURNS_RETAINED attributes
<a href="https://bugs.webkit.org/show_bug.cgi?id=256311">https://bugs.webkit.org/show_bug.cgi?id=256311</a>
&lt;rdar://108895655&gt;

Reviewed by Alex Christensen and Darin Adler.

Use template specialization with SFINAE to add the
CF_RETAINS_RETAINED attribute for CFType objects and the
NS_RETURNS_RETAINED attribute for Cocoa objects.

This allows the clang static analyzer to reason about the
leakRef() method.

* Source/WTF/wtf/RetainPtr.h:
(WTF::RetainPtr&lt;T&gt;::leakRef):
- Use a technique to allow for template specialization without
  using the class template type.
- Note that this specialization must also cover the case of
  using `id` in C++ code.

Canonical link: <a href="https://commits.webkit.org/263742@main">https://commits.webkit.org/263742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3fa3cb7ac49b2e2bb527860ce4dd4d1d2839278

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5992 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7200 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/5634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6025 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5774 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7724 "1 flakes 104 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5753 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5803 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5061 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7243 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3280 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5088 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/12375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/4698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5149 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5164 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/6992 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5232 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5597 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5759 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5045 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1410 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/1334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9156 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5921 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/645 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5406 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1530 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/network/cf/AuthenticationChallenge.h:
```diff
@@ -36,10 +36,16 @@ namespace WebCore {
 class AuthenticationChallenge : public AuthenticationChallengeBase {
 public:
     AuthenticationChallenge() {}
+    WEBCORE_EXPORT AuthenticationChallenge(const AuthenticationChallenge&);
```
These changes are surprising but seem fine.  Why does this structure require a change but others don't?

ddkilzer replied:
From the commit log message:

```
* Source/WebCore/platform/network/cf/AuthenticationChallenge.h:
* Source/WebCore/platform/network/mac/AuthenticationMac.mm:
(WebCore::AuthenticationChallenge::AuthenticationChallenge): Add.
(WebCore::AuthenticationChallenge::operator=): Add.
- Implement the "Rule of Five" for AuthenticationChallenge.  The
  change to leakRef() caused build failures when implicit move
  constructors and copy assignment operators were used in C++
  code because the instance variables were RetainPtr<> objects
  holding Cocoa types.
```

ddkilzer replied:
Interestingly, the [ios](https://ews-build.webkit.org/#/builders/26/builds/5573) and [ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5802) build failures appear to be due to lack of support for `RetainPtr<id>` in pure C++ code.

I guess that would use `CF_RETURNS_RETAINED` because it basically can't use `NS_RETURNS_RETAINED`.

ddkilzer replied:
And it seems that since I implemented a _proper_ move constructor/move assignment operator that empty the values in the moved-from object, I may have triggered some latent use-after-move bugs based on the [mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4962l) crashes.

darinadler reviewed (changes requested):
We should make sure that the move and copy constructors and assignment operators for `RetainPtr` compile and work properly even when the `RetainPtr` is for an Objective-C class and it’s being compiled in a non-Objective-C C++ source file. We should not work around that by changing `AuthenticationChallenge`.

If `leakRef` is causing a problem then we should write the move and copy constructor and assignment operators differently, or enhance `leakRef`.

darinadler commented on Source/WTF/wtf/RetainPtr.h:
```diff
@@ -99,7 +99,27 @@ template<typename T> class RetainPtr {
+    }
+#else
+    template<typename U = T>
+    std::enable_if_t<std::is_same_v<U, id>, PtrType> leakRef() WARN_UNUSED_RETURN {
```
Can we use one of the `RETURNS_RETAINED` macros here? Because this does return something retained.

ddkilzer replied:
I thought neither worked, but I must have been using an earlier iteration of the patch that had a different issue.

`CF_RETURNS_RETAINED` appears to work.  (`NS_RETURNS_RETAINED` won't work because that is only defined in Foundation headers, which aren't available in C++-only sources.)

Commit:
RetainPtr<>::leakRef() should have NS/CF_RETURNS_RETAINED attributes
https://bugs.webkit.org/show_bug.cgi?id=256311
<rdar://108895655>

Reviewed by Alex Christensen and Darin Adler.

Use template specialization with SFINAE to add the
CF_RETAINS_RETAINED attribute for CFType objects and the
NS_RETURNS_RETAINED attribute for Cocoa objects.

This allows the clang static analyzer to reason about the
leakRef() method.

* Source/WTF/wtf/RetainPtr.h:
(WTF::RetainPtr<T>::leakRef):
- Use a technique to allow for template specialization without
  using the class template type.
- Note that this specialization must also cover the case of
  using `id` in C++ code.

Canonical link: https://commits.webkit.org/263742@main
